[APPC-1161] Implemented pre selected payment methods

carvalho2707 opened the pull request:
**What does this PR do?**

   This PR introduces a feature to have pre selected payment methods. There was also a change to payments's flow where the wallet addres(footer) was removed.


**Database changed?**

   No

**Where should the reviewer start?**

- [ ] PaymentMethodsFragment.kt

**How should this be manually tested?**

Until the fix for the orderreferenceid is fixed , change the field orderReference in the RemoteRepository to null.

  Flow on how to test this or QA Tickets related to this use-case: [APPC-1161](https://aptoide.atlassian.net/browse/APPC-1161)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1161](https://aptoide.atlassian.net/browse/APPC-1161)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.)




**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/billing/BillingService.kt:
```diff
@@ -0,0 +1,25 @@
+                       orderReference: String?,
+                       appPackageName: String): Observable<AdyenAuthorization>
+
+  fun getAuthorization(origin: String, priceValue: BigDecimal,
```
These 2 methods can be better splitted (a line can take more arguments before it breaks to another line)

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/billing/adyen/Adyen.java:
```diff
@@ -169,6 +169,12 @@ private void cancelPayment() {
     paymentRequest = null;
   }
 
+  public Completable deletePaymentMethod(PaymentMethod paymentMethod) {
```
paymentMethod is not being used

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java:
```diff
@@ -105,16 +109,22 @@
   private PublishRelay<Boolean> keyboardBuyRelay;
```
walletAddressFooter is not being used

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java:
```diff
@@ -182,14 +198,26 @@ public static AdyenAuthorizationFragment newInstance(String skuId, String type,
+        .getParent();
+    lp = (ViewGroup.MarginLayoutParams) cvvParent.getLayoutParams();
+    lp.setMargins(0, marginTop, 0, 0);
+    cvvParent.setLayoutParams(lp);
```
This could be extracted to a private method, so that in the future people understand better what it does (despite the comment) and to improve readability of the onViewCreated method

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationPresenter.java:
```diff
@@ -128,15 +120,29 @@ public void present(@Nullable Bundle savedInstanceState) {
+          .observeOn(viewScheduler)
+          .doOnNext(o -> {
+            inAppPurchaseInteractor.removePreSelectedPaymentMethod();
+            showMoreMethods();
```
why not just view.showMoreMethods()

carvalho2707 replied:
Because it means we don't have to have a pre selected payment method anymore.

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java:
```diff
@@ -261,12 +266,15 @@ private boolean shouldSwap(List<PaymentMethod> paymentMethods, int position) {
+            .getName() == (Gateway.Name.myappcoins)) {
+          paymentMethod.getAvailability();
+          if (paymentMethod.getAvailability()
+              .equals("UNAVAILABLE")) {
```
Why did you change the logic here?

carvalho2707 replied:
Because paymentMethod.getAvailability() returns non null values always and thus the check is unnecessary

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java:
```diff
@@ -284,4 +292,31 @@ private PaymentMethod mapPaymentMethods(PaymentMethodEntity paymentMethod,
+  void savePreSelectedPaymentMethod(String paymentMethod) {
+    SharedPreferences.Editor editor = sharedPreferences.edit();
+    editor.putString(PRE_SELECTED_PAYMENT_METHOD_KEY, paymentMethod);
+    editor.putString(LAST_USED_PAYMENT_METHOD_KEY, paymentMethod);
```
Isn’t the last always the same as the pre-selected?

carvalho2707 replied:
PRE_SELECTED_PAYMENT_METHOD_KEY is for the new flow where we don't show the payment methods list.
LAST_USED_PAYMENT_METHOD_KEY is used when there is no pre selected payment method but we want to have the radio check on the last used (before clicking on "more payment method)

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java:
```diff
@@ -68,18 +71,17 @@
-  @Inject WalletService walletService;
   @Inject GamificationInteractor gamification;
   @Inject PaymentMethodsMapper paymentMethodsMapper;
+  @Inject BillingFactory billingFactory;
```
Not used

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt:
```diff
@@ -0,0 +1,340 @@
+    private val transaction: TransactionBuilder,
+    private val paymentMethodsMapper: PaymentMethodsMapper) {
+
+  private var transactionValue: Double = 0.toDouble()
```
This can passed as an argument of the presenter

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt:
```diff
@@ -0,0 +1,337 @@
+  }
+
+  private fun getLastUsedPaymentMethod(paymentMethods: List<PaymentMethod>): String {
+    val preSelectedPreference = inAppPurchaseInteractor.lastUsedPaymentMethod
```
variable name

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt:
```diff
@@ -0,0 +1,337 @@
+            return@flatMapCompletable getPurchases(transaction.skuId).doOnNext { purchase ->
+              finish(purchase, true)
+            }
+                .ignoreElements()
```
Use “it” instead of purchase. Remove line break of ignore elements

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt:
```diff
@@ -0,0 +1,337 @@
+        .subscribe({ }, { this.showError(it) }))
+  }
+
+  private fun finish(purchase: Purchase, itemAlreadyOwned: Boolean?) {
```
itemAlreadyOwned is never null

JDSM01 commented on app/src/main/res/layout/selected_payment_method_cc.xml:
```diff
@@ -0,0 +1,69 @@
+      android:checked="true"
+      android:text="@string/dialog_credit_card_remember"
+      android:textSize="11sp"
+      android:visibility="visible"
```
not needed

JDSM01 commented on app/src/main/res/layout-land/dialog_credit_card_authorization_pre_selected.xml:
```diff
@@ -0,0 +1,191 @@
+        android:layout_height="80dp"
+        android:layout_centerInParent="true"
+        android:indeterminateDrawable="@drawable/gradient_progress"
+        android:visibility="visible"
```
not needed

JDSM01 commented on app/src/main/res/layout-land/payment_methods_layout.xml:
```diff
@@ -278,8 +184,38 @@
+
+      <TextView
+          android:id="@+id/more_payment_methods"
+          style="@style/TextAppearance.AppCompat.Body1.Roboto.Medium"
```
Text color doesn’t match the one in zeplin

JDSM01 commented on app/src/main/res/layout/selected_payment_method.xml:
```diff
@@ -0,0 +1,54 @@
+      />
+
+  <TextView
+      android:id="@+id/payment_method_description"
```
Text is not centered in some cases in portrait (e.g. PayPal)

JDSM01 commented on app/src/main/res/layout/selected_payment_method.xml:
```diff
@@ -0,0 +1,54 @@
+      tools:text="Paypal"
+      />
+  <TextView
+      android:id="@+id/payment_method_secondary"
```
This could be gone by default

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/IabActivity.kt:
```diff
@@ -0,0 +1,253 @@
+
+    @JvmStatic
+    fun newIntent(activity: Activity, previousIntent: Intent, transaction: TransactionBuilder,
+                  isBds: Boolean?, developerPayload: String): Intent {
```
developerPayload can be null

JDSM01 reviewed (changes requested):
Margin from header is to far on the left
Ask a friend not working
PaymentMethodView -> rotate -> buy with credits -> previous view doesn’t disappear
Landscape margin bottom in pre-selected view is too large
Portrait margin bottom in pre-selected view is too short
Pre-selectedView -> rotate-> rotate -> paymentMethodView appears in the background
Going from a  payment with bonus and a payment without bonus makes view shrink/enlarge
Use case missing: buy with credits --> no longer have credits --> should show payment list view
Item already owned not working

carvalho2707 committed:
[APPC-1161] Implemented pre selected payment methods

carvalho2707 committed:
[APPC-1161] Rebased conflicts fix from pre selected payments on top of dev

carvalho2707 committed:
[APPC-1161] Fixed bonus layout for cc dialog land

JDSM01 reviewed (changes requested):
- [x] Bonus landscape layout is still botched

- [x] Payment List layout doesn’t match the one in the merged appcoins branch. (Space to the top, payment_methods_layout)

- [x] Bottom list line (top bonus line) doesn’t appear in portrait ( if preselected and then click on more)

- [x] ProductName in IabActivity can be null

- [x] Item already owned nao tem listener no botao de erro

- [x] Margin between bonus layout and bottom separator(botto top line)

carvalho2707 committed:
[APPC-1161] ui improvements after review comments

carvalho2707 committed:
[APPC-1161] Added margin top to payments methods layout landscape

carvalho2707 committed:
[APPC-1161] Fixed selected payment method layout

carvalho2707 committed:
[APPC-1161] Removed empty space between bottom separator and pre selected payment method

carvalho2707 committed:
[APPC-1161] Fixed wrong icon url for appc creddits in pre selected payment methods

carvalho2707 committed:
[APPC-1161] Fixed dialog credit card header height

JDSM01 reviewed (changes requested):
Wrong flow: Made a buy with credit card, mock response to remove CC from response, Merge Appcoins appears as pre-selected

carvalho2707 committed:
Merge branch 'dev' into feature/APPC-1161-pre-selected-previously-used-payment-method

carvalho2707 committed:
[APPC-1161] Fixed pre selected payment method when it was missing in the response

carvalho2707 committed:
Merge branch 'dev' into feature/APPC-1161-pre-selected-previously-used-payment-method

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java:
```diff
@@ -211,7 +211,8 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
     preSelectedIcon = preSelectedMethodView.findViewById(R.id.payment_method_ic);
     preSelectedName = preSelectedMethodView.findViewById(R.id.payment_method_description);
-    pareSelectedNameSingle = preSelectedMethodView.findViewById(R.id.payment_method_description_single);
+    pareSelectedNameSingle =
```
name

carvalho2707 committed:
[APPC-1161] Changed variable name

JDSM01 reviewed (changes requested):
- [x] Local Payment asynchronous

- [x] Buy button doesn’t work on pre-selected

- [x] Buy on CC without cvc gives an error that ends the flow

carvalho2707 committed:
Local Payment asynchronous

  Buy button doesn’t work on pre-selected

  Buy on CC without cvc gives an error that ends the flow
